openmeter/customer/event.go

```diff
@@ -0,0 +1,145 @@
+package customer
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/openmeterio/openmeter/openmeter/event/metadata"
+)
+
+const (
+	CustomerEventSubsystem  metadata.EventSubsystem = "customer"
+	CustomerCreateEventName metadata.EventName      = "customer.created"
+	CustomerUpdateEventName metadata.EventName      = "customer.updated"
+	CustomerDeleteEventName metadata.EventName      = "customer.deleted"
+)
+
+// NewCustomerCreateEvent creates a new customer create event
+func NewCustomerCreateEvent(customer Customer) CustomerCreateEvent {
+	return CustomerCreateEvent{
+		EventEntityMutationPayload: metadata.EventEntityMutationPayload[Customer]{
+			Entity:       metadata.EntityCustomer,
+			MutationType: metadata.EntityMutationTypeCreate,
+			New:          &customer,
+		},
+	}
+}
+
+// CustomerCreateEvent is an event that is emitted when a customer is created
+type CustomerCreateEvent struct {
+	metadata.EventEntityMutationPayload[Customer] `json:",inline"`
+}
+
+func (e CustomerCreateEvent) EventName() string {
+	return metadata.GetEventName(metadata.EventType{
+		Subsystem: CustomerEventSubsystem,
+		Name:      CustomerCreateEventName,
+		Version:   "v1",
+	})
+}
+
+func (e CustomerCreateEvent) EventMetadata() metadata.EventMetadata {
+	return metadata.EventMetadata{
+		ID:     e.New.ID,
+		Source: metadata.ComposeResourcePath(e.New.Namespace, metadata.EntityCustomer, e.New.ID),
+		Time:   e.New.CreatedAt,
+	}
+}
+
+func (e CustomerCreateEvent) Validate() error {
+	var errs []error
+
+	if err := e.EventEntityMutationPayload.Validate(); err != nil {
+		errs = append(errs, fmt.Errorf("customer create event: %w", err))
+	}
+
+	return errors.Join(errs...)
+}
+
+// NewCustomerUpdateEvent creates a new customer update event
+func NewCustomerUpdateEvent(new Customer, previous Customer) CustomerUpdateEvent {
+	return CustomerUpdateEvent{
+		EventEntityMutationPayload: metadata.EventEntityMutationPayload[Customer]{
+			Entity:       metadata.EntityCustomer,
+			MutationType: metadata.EntityMutationTypeUpdate,
+			New:          &new,
+			Previous:     &previous,
+		},
+	}
+}
+
+// CustomerUpdateEvent is an event that is emitted when a customer is updated
+type CustomerUpdateEvent struct {
+	metadata.EventEntityMutationPayload[Customer] `json:",inline"`
+}
+
+func (e CustomerUpdateEvent) EventName() string {
+	return metadata.GetEventName(metadata.EventType{
+		Subsystem: CustomerEventSubsystem,
+		Name:      CustomerUpdateEventName,
+		Version:   "v1",
+	})
+}
+
+func (e CustomerUpdateEvent) EventMetadata() metadata.EventMetadata {
+	return metadata.EventMetadata{
+		ID:     metadata.GetMutationEventID(metadata.EntityMutationTypeUpdate, e.Previous.ID, e.New.UpdatedAt),
```

Comment:
This is autogenerated as an ULID, why do we need this to be said. Is something downstream depends on this?

https://github.com/openmeterio/openmeter/blob/3931852e80aad734048eb6fad89e05927957649d/openmeter/watermill/marshaler/marshaler.go#L103